overlord/devicestate/devicestate.go

```diff
@@ -478,25 +482,76 @@ func remodelTasks(ctx context.Context, st *state.State, current, new *asserts.Mo
 		if firstInstallInChain == nil {
 			firstInstallInChain = installFirst
 		}
-	}
-	// Make sure the first install waits for the last download. With this
-	// our (simplified) wait chain looks like:
-	// download1 <- verify1 <- download2 <- verify2 <- download3 <- verify3 <- install1 <- install2 <- install3
+		// download is always a first task of the 'download' phase
+		snapSetupTasks = append(snapSetupTasks, downloadLast.ID())
+	}
+	// Make sure the first install waits for the recovery system (only in
+	// UC20) which waits for the last download. With this our (simplified)
+	// wait chain looks like this:
+	//
+	// download1
+	//   ^- verify1
+	//        ^- download2
+	//             ^- verify2
+	//                  ^- download3
+	//                       ^- verify3
+	//                            ^- recovery (UC20)
+	//                                 ^- install1
+	//                                      ^- install2
+	//                                           ^- install3
 	if firstInstallInChain != nil && lastDownloadInChain != nil {
 		firstInstallInChain.WaitFor(lastDownloadInChain)
 	}
 
+	recoverySetupTaskID := ""
+	if current.Grade() != asserts.ModelGradeUnset && new.Grade() != asserts.ModelGradeUnset {
+		// right now it's only possible to remodel from a UC20 system to
+		// another UC20 system, when doing so create a recovery
+		label := timeNow().Format("20060102")
+		createRecoveryTasks, err := createRecoverySystemTasks(st, label, snapSetupTasks)
+		if err != nil {
+			return nil, err
+		}
+		if lastDownloadInChain != nil {
+			// wait for all snaps that need to be downloaded
+			createRecoveryTasks.WaitFor(lastDownloadInChain)
+		}
+		if firstInstallInChain != nil {
+			// when any snap installations need to happen, they
+			// should also wait for recovery system to be created
+			firstInstallInChain.WaitAll(createRecoveryTasks)
+		}
+		tss = append(tss, createRecoveryTasks)
+		recoverySetupTaskID = createRecoveryTasks.Tasks()[0].ID()
+	}
+
 	// Set the new model assertion - this *must* be the last thing done
 	// by the change.
 	setModel := st.NewTask("set-model", i18n.G("Set new model assertion"))
 	for _, tsPrev := range tss {
 		setModel.WaitAll(tsPrev)
 	}
+	if recoverySetupTaskID != "" {
+		// set model needs to access information about the recovery
+		// system
+		setModel.Set("recovery-system-setup-task", recoverySetupTaskID)
+	}
 	tss = append(tss, state.NewTaskSet(setModel))
 
 	return tss, nil
 }
 
+var allowUC20RemodelTesting = false
+
+func AllowUC20RemodelTesting(allow bool) (restore func()) {
```

Comment:
thanks for doing it this way, but I think this probably merits a comment

// XXX AllowUC20RemodelTesting is a temporary helper to allow testing remodeling of UC20 before the implementation is complete and the policy for this settled,  will go away.

Reply:
added